### PR TITLE
CST-316 AIDA Revised FOV Definition

### PR DESCRIPTION
**Pull request by cpparts**
AIDA viewport fov was redefined to refer to the short side of the viewport  in https://github.com/astropy/astro-image-display-api/pull/56: 
"The field of view (FOV) of the viewport. If not given, the current FOV is used. If a float is given, it is interpreted as a size in pixels. For viewers that are not square, the FOV is interpreted as the size of the shorter side of the viewer such that the FOV is guaranteed to be entirely visible regardless of the aspect ratio of the viewer."

This PR updates the AIDA `get_viewport` and `set_viewport` to use this new re-definition. Since the method relies on knowing the parameter `fov_xy` to determine which side is the short side of the viewport, we have been forced to comment out much of the getting and setting testing previously implemented. Setting of `fov_xy` depends on returns from aladin-lite, which is not possible to test in python under our current setup. For those interested in manually confirming functionality and exploring this new implementation, I have included a notebook that demonstrates a roundtrip getting and setting of the viewport.

### CODE REVIEW

File: mast_aladin_lite/aida.py [hunk 1]
```diff
@@ -26,6 +27,8 @@ def set_viewport(self, center):
         ----------
         center : `~astropy.coordinates.SkyCoord`
             Center the viewer on this coordinate.
+        fov : `~astropy.coordinates.Angle` or float
+            Set the width of the viewport to span `field_of_view`.
```
Comment (bmorris3) [inline mast_aladin_lite/aida.py:31]:
```suggestion
            Set the length of the viewport's smaller axis to span `fov`, as a
            `~astropy.coordinates.Angle` or as a float in units of degrees.
```

File: mast_aladin_lite/aida.py [hunk 2]
```diff
@@ -34,12 +37,38 @@ def set_viewport(self, center):
...
+                if fov < 1:
+                    scale_factor = fov
 
-        self.app.target = center
+            else:
+                raise ValueError(
+                    f"`fov` must be an `~astropy.coordinates.Angle` or float, got {fov=}"
+                )
+
+            current_fov = self.app.fov.value
```
Comment (bmorris3) [inline mast_aladin_lite/aida.py:62]: To make this safer and clearer:
```suggestion
            current_fov = self.app.fov.to_value(u.deg)
```

File: mast_aladin_lite/aida.py [hunk 3]
```diff
@@ -34,12 +37,38 @@ def set_viewport(self, center):
...
+            aspect_ratio = float(self.app._fov_xy["y"]/self.app._fov_xy["x"])
+
+            if scale_factor == 0:
```
Comment (bmorris3) [inline mast_aladin_lite/aida.py:65]: This `if` is always True, right? Maybe we don't need to set `scale_factor = 0` earlier?

File: notebooks/AIDA_fov_getting_and_setting.ipynb [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "aadd0c4f-f197-4f2e-8d14-871f88e74537",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# retrieve final fov and fov_xy, the x and y fov initial and final should match, so check that\n",
+    "final_fov = mast_aladin.aid.get_viewport()[\"fov\"]\n",
```
[the remark this answers is not in the document]
Comment (bmorris3) [reply notebooks/AIDA_fov_getting_and_setting.ipynb:114]: Aha! The issue here is that I didn't have the cobalt branch of ipyaladin installed at the time.

File: mast_aladin_lite/aida.py (hunk 1 above)
Comment (cpparts) [reply mast_aladin_lite/aida.py:31]: The one thing that might complicate this is that the float can be in terms of degrees or it can accept a decimal to change the fov by that factor. Do we want to not support that?

File: notebooks/AIDA_fov_getting_and_setting.ipynb [hunk 5]
```diff
@@ -0,0 +1,154 @@
+{
+ "cells": [
+  {
```
[the remark this answers is not in the document]
Comment (cpparts) [reply notebooks/AIDA_fov_getting_and_setting.ipynb:3]: This should be handled by installing the cobalt dev branch that we have in the dependencies from this [commit](https://github.com/spacetelescope/mast-aladin-lite/commit/d33ceb501e5e5b6146bae03f3ccca813eab9973e). Do we want to have each notebook include an install statement for this branch until we can rely on the regular ipyaladin?

File: mast_aladin_lite/aida.py (hunk 1 above)
Comment (bmorris3) [reply mast_aladin_lite/aida.py:31]: Oh, no we don't want to support that.

File: mast_aladin_lite/aida.py [hunk 6]
```diff
@@ -34,12 +38,35 @@ def set_viewport(self, center):
...
+        if center is not None:
+            if not isinstance(center, SkyCoord):
+                raise TypeError(
+                    "`center` must be a SkyCoord object."
+                )
+
+            self.app.target = center
+
+        if fov is not None:
+            scale_factor = 0
```
Comment (bmorris3) [inline mast_aladin_lite/aida.py:50]: Isn't `scale_factor` always updated below? I'm not sure it needs to be here.

**Review by bmorris3 — COMMENTED** [review]
Getting so close! I'm relieved to _not_ see the 'float < 1 is relative zoom' case.

File: mast_aladin_lite/aida.py (hunk 6 above)
Comment (tomdonaldson) [reply mast_aladin_lite/aida.py:50]: But you're right that it's unnecessary.  If the test case cover all the edge cases, then removing it would be clearer since there could be no test for using the default.

File: notebooks/AIDA_fov_getting_and_setting.ipynb [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "64ece6f1-e55e-4b77-9a99-6b9fe6931cbe",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from mast_aladin_lite import MastAladin\n",
+    "from astropy.coordinates import SkyCoord, Angle\n",
+    "import astropy.units as u\n",
+    "from astropy.tests.helper import assert_quantity_allclose\n",
```
Comment (tomdonaldson) [inline notebooks/AIDA_fov_getting_and_setting.ipynb:25]: This doesn't seem to be used and requires me to have the astropy testing dependencies.

Comment (cpparts) [reply notebooks/AIDA_fov_getting_and_setting.ipynb:25]: initially I used them when playing around with all of the viewport args, but I missed removing this when I removed that portion of the notebook

File: mast_aladin_lite/aida.py (hunk 6 above)
Comment (cpparts) [reply mast_aladin_lite/aida.py:50]: I can remove this! The test case rn is the round-trip notebook since we can't retrieve the fov_xy from aladin-lite

File: notebooks/AIDA_fov_getting_and_setting.ipynb (hunk 7 above)
Comment (cpparts) [reply notebooks/AIDA_fov_getting_and_setting.ipynb:25]: These are used in the test_aida_example.py notebook, which is where I moved this code from when I realized there was no way to actually test getting fov_xy values. I removed them for now since they are not needed for this notebook at this point, but it's definitely something to revisit as we introduce more robust testing!

File: mast_aladin_lite/aida.py [hunk 8]
```diff
@@ -34,12 +38,35 @@ def set_viewport(self, center):
...
+                fov = fov.value
+
+            elif not isinstance(fov, (float, int)):
+                raise ValueError(
+                    f"`fov` must be an `~astropy.coordinates.Angle` or float, got {fov=}"
+                )
 
-        self.app.target = center
+            current_fov = self.app.fov.to_value(u.deg)
+            aspect_ratio = float(self.app._fov_xy["y"] / self.app._fov_xy["x"])
```
Comment (tomdonaldson) [inline mast_aladin_lite/aida.py:59]: I had to look this up since different languages do different things, but I think these `float()` calls are redundant because Python division always returns a float.    
https://docs.python.org/3/library/operator.html#mapping-operators-to-functions
https://docs.python.org/3/library/operator.html#mapping-operators-to-functions

I was checking because if the division could have returned a rounded-off int when dividing 2 ints, then this `float()` should be on one (or both) of the operands instead of on the result which would already have been rounded.
